nftables: split python package from main

alaviss opened the pull request:
## Description of changes

Starting nftables 1.0.9, the python bindings are no longer built and installed by the autotools build system and instead must be installed manually.

Related nftables commit: https://git.netfilter.org/nftables/commit/configure.ac?id=b3def33efecb2f7be39fc9aefc9546907202056c

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

alaviss committed:
nftables: split python package from main

Starting nftables 1.0.9, the python bindings are no longer built and
installed by the autotools build system and instead must be installed
manually.

Co-authored-by: ajs124 <ajs124@users.noreply.github.com>

ajs124 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -8254,10 +8254,7 @@ self: super: with self; {
-    python3 = python;
-    withPython = true;
-  });
+  nftables = callPackage ../os-specific/linux/nftables/python.nix { nftables = pkgs.nftables; };
```
```suggestion
  nftables = callPackage ../os-specific/linux/nftables/python.nix {
    inherit (pkgs) nftables;
  };
```
seems to be the more common pattern in this file

ajs124 reviewed (commented):
diff LGTM, but didn't test

PABLEXWorld commented:
This PR fixes issue 293778

quentinmit commented on pkgs/os-specific/linux/nftables/python.nix:
```diff
@@ -0,0 +1,26 @@
+
+buildPythonPackage {
+  pname = "nftables";
+  inherit (nftables) version src meta;
```
You can't just inherit `meta` because you'll get `meta.mainProgram`, which isn't correct for the Python package. Instead I did

```nix
  meta = {
    inherit (nftables.meta) description homepage license platforms maintainers;
  };
```

quentinmit commented on pkgs/os-specific/linux/nftables/python.nix:
```diff
@@ -0,0 +1,26 @@
+  pname = "nftables";
+  inherit (nftables) version src meta;
+
+  patches = [ ./fix-py-libnftables.patch ];
```
Why bother with a patch when you can just `--replace libnftables.so.1` directly in the postPatch?

alaviss replied:
`--replace` is silent when it fails, a patch will fail to apply if this ever change.

Izorkin replied:
And if replace it with `--replace-fail`?

alaviss replied:
TIL, updated, thanks.

Izorkin replied:
Needs to delete the file with the patch.

quentinmit commented on pkgs/os-specific/linux/nftables/python.nix:
```diff
@@ -0,0 +1,26 @@
+  '';
+
+  preConfigure = ''
+    cd py
```
I believe this is what `setSourceRoot` is for:

```nix
setSourceRoot = "sourceRoot=$(echo */py)";
```

alaviss committed:
nftables: remove python patch

Replaced the patch with `--replace-fail` based on @quentinmit and
@Izorkin recommendation.

Izorkin commented:
@alaviss it's probably worth squashing last 2 commits into one.

SuperSandro2000 commented on pkgs/os-specific/linux/nftables/python.nix:
```diff
@@ -0,0 +1,26 @@
+  pyproject = true;
+  pythonImportsCheck = [ "nftables" ];
+
+  meta = {
```
We should probably just inherit the entire meta here

alaviss replied:
`nftables.meta` contains `mainProgram`, which is not a thing for this package.

SuperSandro2000 commented on pkgs/os-specific/linux/nftables/python.nix:
```diff
@@ -0,0 +1,26 @@
+
+  nativeBuildInputs = [ setuptools ];
+
+  pyproject = true;
```
We usually write pyproject straight after version

SuperSandro2000 commented:
@ofborg build nftables
